## Q1. Write a function that given your birthday (as a date), returns how old you are in years.

In [90]:
from datetime import date

def age(birthDate):
  today=date.today()
  age=today.year - birthDate.year - ((today.month, today.day) < (birthDate.month, birthDate.day))
  return age

print(age(date(1999,3,20)),'years')          #format: YY/MM/DD

22 years


## Q2. What happens if you parse a string that contains invalid dates?
 
  Ex: ymd(c("2010-10-10", "bananas")

In [91]:
from datetime import datetime
from datetime import timedelta

Begindatestring='2010-10-10'
Begindate = datetime.strptime(Begindatestring, "%Y-%m-%d")

enddate=Begindate + timedelta(days='Banana')
print(enddate)

TypeError: ignored

If number of days is int it will get added to the begindate but if it is string the error message in python will be: unsupported type for timedelta days component: str

## Q3. How do you import default dataset presend in R to work session.

In [92]:
# there are many libraries in python 
from sklearn.datasets import load_breast_cancer
import pandas as pd

data=load_breast_cancer()
df=pd.DataFrame(data.data, columns=data.feature_names)
df.head()

mean radius  mean texture  ...  worst symmetry  worst fractal dimension
0        17.99         10.38  ...          0.4601                  0.11890
1        20.57         17.77  ...          0.2750                  0.08902
2        19.69         21.25  ...          0.3613                  0.08758
3        11.42         20.38  ...          0.6638                  0.17300
4        20.29         14.34  ...          0.2364                  0.07678

[5 rows x 30 columns]

Using the flights data, Find all the flights that.
 Instruction: Paste just the query which gives the correct output.

In [93]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [94]:
link='/content/drive/MyDrive/flight_data.csv'

In [95]:
import pandas as pd
data=pd.read_csv(link)
data.head()

year  month  day  dep_time  ...  distance  hour  minute      time_hour
0  2013      1    1     517.0  ...      1400     5      15  1/1/2013 5:00
1  2013      1    1     533.0  ...      1416     5      29  1/1/2013 5:00
2  2013      1    1     542.0  ...      1089     5      40  1/1/2013 5:00
3  2013      1    1     544.0  ...      1576     5      45  1/1/2013 5:00
4  2013      1    1     554.0  ...       762     6       0  1/1/2013 6:00

[5 rows x 19 columns]

In [96]:
data.isna().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

removing all null values

In [97]:
data=data.dropna()

In [98]:
data.isna().sum()

year              0
month             0
day               0
dep_time          0
sched_dep_time    0
dep_delay         0
arr_time          0
sched_arr_time    0
arr_delay         0
carrier           0
flight            0
tailnum           0
origin            0
dest              0
air_time          0
distance          0
hour              0
minute            0
time_hour         0
dtype: int64

In [99]:
len(data)

327346

## Q4 Had an delay in arrival of 3 or less hours.

In [100]:
# arrival delap less than or equal ro 3
len(data[data['arr_delay'] <= 3]) 

208858

In [101]:
# departure delay less than or equal to 3
len(data[data['dep_delay'] <= 3])

219272

## Q5. Flew to Houston (IAH or HOU).

In [102]:
len(data[data['dest']=='IAH'])

7085

Total 7085 flights flew to houston.

## Q6. Were operated by United, America or delta.

Consedring AA as American, US as United and DL as delta.

In [103]:
len(data[((data.carrier=='AA')|(data.carrier=='UA')|(data.carrier=='DL'))])

137387

## Q7. Departed in summer (July, August, and September).

July is 7

August is 8

September is 9

In [104]:
len(data[((data.month==7)| (data.month == 8) | (data.month ==9))])

84059

## Q8. Arrived more than two hours late, but didn't leave late.

In [122]:

dd=data[data.arr_delay>120.0]
len(dd[dd.dep_delay<=0])

29

In [131]:
(data[data.arr_delay> 120.0].dep_delay <=0).sum()

29

29 flights arrived more than 120 min late but departed at right time or earlier.

## Q9. Were delayed by at least an hour, but made up over 30 minutes in flight

In [143]:
mean_speed=data['distance'].mean()/data['air_time'].mean()
print(mean_speed)

6.957302681047817


In [151]:
(data[(data['distance']/mean_speed) > (data['air_time']+30.0)].dep_delay >= 30.0).sum()


3493

Used mean time to calculate the average time taken by flight to fly each km and them multiplied with total ditance to fing the real air time it should take. There were 3493 flights which departed 30 min late but made up in the flight time.

## Q10. Departed between midnight and 6am (inclusive)

In [137]:

len(data[data['sched_dep_time']<=600.0])

8776

8776 # of flights departed between midnight and 6am